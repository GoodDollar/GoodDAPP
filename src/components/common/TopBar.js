import React from 'react'

import GDStore from '../../lib/undux/GDStore'
import Avatar from './Avatar'
import BigGoodDollar from './BigGoodDollar'
import Section from './Section'

const TopBar = ({ hideBalance, push }) => {
  const store = GDStore.useStore()
  const { balance } = store.get('account')
  const { avatar } = store.get('profile')

  return (
    <Section>
      <Section.Row>
        <Avatar source={avatar} onPress={push && (() => push('Profile'))} />
        {!hideBalance && <BigGoodDollar number={balance} />}
      </Section.Row>
    </Section>
  )
}

export default TopBar
