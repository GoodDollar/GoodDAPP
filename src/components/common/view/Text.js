// @flow
import React from 'react'
import { StyleSheet } from 'react-native'
import { Text as PaperText } from 'react-native-paper'

import normalize from '../../../lib/utils/normalizeText'
import { withStyles } from '../../../lib/styles'
import { calculateFontFamily, calculateFontWeight } from '../../../lib/utils/fonts'

const LINE_HEIGHT_FACTOR = 1.2

class Text extends React.Component {
  _root

  /**
   * @internal
   * Required for when the component is wrapped inside a TouchableOpacity* element
   * https://facebook.github.io/react-native/docs/direct-manipulation
   */
  setNativeProps(...args) {
    return this._root && this._root.setNativeProps(...args)
  }

  render() {
    const {
      style,
      styles,
      theme,
      color,
      textAlign,
      fontWeight,
      fontFamily,
      fontSize,
      lineHeight,
      textDecorationLine,
      textTransform,
      ...rest
    } = this.props

    return <PaperText {...rest} ref={c => (this._root = c)} style={[styles.text, style]} />
  }
}

/**
 * Returns the linespacing associated to the font-size based on guidelines
 * Defaults to 1em if size is not in the list
 * @param {number} fontSize
 * @returns {*|string}
 */
const relatedLineSpacing = fontSize =>
  ({
    8: 14,
    10: 14,
    12: 16,
    14: 20,
    16: 22,
    18: 24,
    22: 30,
    24: 30,
    36: 30,
    42: 30,
  }[fontSize] || fontSize * LINE_HEIGHT_FACTOR)

const getStylesFromProps = ({
  theme,
  color,
  textAlign,
  style,
  fontWeight,
  fontFamily,
  fontSize,
  lineHeight,
  textDecorationLine,
  textTransform,
  letterSpacing,
  normalizeText = true,
}) => {
  const calculatedFontSize = Number.isFinite(fontSize) ? fontSize : 16
  const calculatedLineHeight = lineHeight || relatedLineSpacing(calculatedFontSize)

  const selectedFontFamily = theme.fonts[fontFamily] || fontFamily
  const selectedFontWeight = StyleSheet.flatten(style)?.fontWeight || fontWeight

  const calculatedFontWeight = isNaN(selectedFontWeight) ? calculateFontWeight(selectedFontWeight) : selectedFontWeight
  const calculatedFontFamily = calculateFontFamily(selectedFontFamily, selectedFontWeight)

  return {
    text: {
      color: theme.colors[color] || color || theme.colors.darkGray,
      textAlign: textAlign || 'center',
      fontWeight: calculatedFontWeight,
      fontFamily: calculatedFontFamily,
      fontSize: normalizeText ? normalize(calculatedFontSize) : calculatedFontSize,
      lineHeight: normalizeText ? normalize(calculatedLineHeight) : calculatedLineHeight,
      textTransform: textTransform || 'none',
      textDecorationLine: textDecorationLine || 'none',
      letterSpacing,
    },
  }
}

export default withStyles(getStylesFromProps)(Text)
