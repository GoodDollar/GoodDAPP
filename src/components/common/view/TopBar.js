import React, { useCallback, useContext } from 'react'
import { Platform, StyleSheet } from 'react-native'
import Section from '../layout/Section'
import useProfile from '../../../lib/userStorage/useProfile'
import { TokenContext } from '../../../lib/wallet/GoodWalletProvider'
import { theme } from '../../theme/styles'
import Config from '../../../config/config'
import Avatar from './Avatar'
import BigGoodDollar from './BigGoodDollar'

/**
 * TopBar - used To display contextual information in a small container
 * @param {object} props - an object with props
 * @param {boolean} props.hideBalance - if falsy balance will be displayed
 * @param {function} props.push - pushes a route to the nav stack. When called, apps navigates to the specified ruote
 * @param {React.Node} props.children
 * @returns {React.Node}
 */
const TopBar = ({
  hideBalance,
  push,
  children,
  style,
  hideProfile = false,
  profileAsLink = false,
  contentStyle,
  avatarSize,
  isBridge,
  network,
}) => {
  const { smallAvatar: avatar } = useProfile()
  const { balance, token, native } = useContext(TokenContext)
  const isNativeToken = Config.isDeltaApp && native

  const redirectToProfile = useCallback(() => {
    if (!push || !profileAsLink) {
      return
    }

    push('Profile')
  }, [push, profileAsLink])

  const onPressAvatar = () => {
    if (push) {
      return profileAsLink ? redirectToProfile : null
    }
  }

  return (
    <Section style={[styles.topBar, style]}>
      <Section.Row
        alignItems={Platform.select({
          web: 'center',
        })}
        style={[contentStyle, { flexDirection: isBridge ? 'row-reverse' : 'row' }]}
      >
        {/*
          if children exist, it will be rendered
          if children=undefined and hideBalance=false, BigGoodDollar will be rendered
          if children=undefined and hideBalance=true, nothing will be rendered
          */}
        <Section.Text style={styles.balance}>
          {!hideBalance && (
            <BigGoodDollar style={styles.bigGoodDollar} number={balance} unit={isNativeToken ? token : undefined} />
          )}
        </Section.Text>

        {children}
        {hideProfile !== true && !isBridge && (!children || hideBalance) && (
          <Avatar
            source={
              avatar //if not already displaying two items show also avatar
            }
            onPress={onPressAvatar}
            size={avatarSize}
          />
        )}
        {isBridge && <Section.Text style={styles.networkName}> {network} G$ </Section.Text>}
      </Section.Row>
    </Section>
  )
}

const styles = StyleSheet.create({
  topBar: {
    justifyContent: 'center',
    marginBottom: 8,
    paddingBottom: 8,
    paddingLeft: 12,
    paddingRight: 8,
    paddingTop: 8,
    height: 62,
  },
  networkName: {
    fontSize: 16,
    lineHeight: 20,
    color: theme.colors.lighterGray,
  },
  balance: {
    ...Platform.select({
      android: { height: 10 },
      default: {},
    }),
  },
  bigGoodDollar: {
    ...Platform.select({
      android: { height: 30 },
      default: {},
    }),
    display: 'flex',
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
  },
})

export default TopBar
