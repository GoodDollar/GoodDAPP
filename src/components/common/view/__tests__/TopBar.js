// eslint-disable-next-line import/order
import React from 'react'
import { View } from 'react-native'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

import { withThemeProvider, withUserStorage } from '../../../../__tests__/__util__'
import TopBar from '../TopBar'
import Text from '../Text'

jest.setTimeout(30000)

describe('TopBar', () => {
  const WrappedTopBar = withThemeProvider(withUserStorage(TopBar))

  it('matches snapshot without balance', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<WrappedTopBar hideBalance />)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches snapshot with balance', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<WrappedTopBar />)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`should render the children component`, async () => {
    // Given
    let component

    await renderer.act(
      // eslint-disable-next-line require-await
      async () =>
        (component = await renderer.create(
          <WrappedTopBar>
            <View>
              <Text>Children element</Text>
            </View>
          </WrappedTopBar>,
        )),
    )

    // When
    const tree = component.toJSON()

    // Then
    expect(tree).toMatchSnapshot()
  })
})
