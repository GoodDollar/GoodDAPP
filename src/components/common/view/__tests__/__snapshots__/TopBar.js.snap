// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar matches snapshot with balance 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-1f0042m r-height-yivrw8 r-justifyContent-1777fci r-marginBottom-5oul0u r-paddingBottom-xd6kpl r-paddingHorizontal-1e081e0 r-paddingLeft-f727ji r-paddingRight-1pyaxff r-paddingTop-tskmnb r-paddingVertical-1yzf0co"
  >
    <div
      className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1d09ksm r-display-6koalj r-flexDirection-18u37iz"
      >
        <div
          className="css-text-901oao r-color-181bzza r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-b88u0q r-lineHeight-qvk6io r-marginRight-a5pmau r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          dir="auto"
        >
          0
        </div>
        <div
          className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
        >
          <div
            className="css-text-901oao r-color-181bzza r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-b88u0q r-lineHeight-qvk6io r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
            dir="auto"
          >
            G
          </div>
          <div
            className="css-text-901oao r-color-181bzza r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-b88u0q r-lineHeight-qvk6io r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
            dir="auto"
          >
            $
          </div>
        </div>
      </div>
      <div
        className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-qt01xk r-borderColor-y68mpx r-borderWidth-rs99b7 r-cursor-1loqt21 r-justifyContent-1777fci r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "borderBottomLeftRadius": "21px",
            "borderBottomRightRadius": "21px",
            "borderTopLeftRadius": "21px",
            "borderTopRightRadius": "21px",
            "height": "42px",
            "transitionDuration": "0s",
            "width": "42px",
          }
        }
        tabIndex="0"
      >
        <div
          className="css-view-1dbjc4n"
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0.00)",
              "height": "42px",
              "width": "42px",
            }
          }
        >
          <unknownProfile.svg />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopBar matches snapshot without balance 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-1f0042m r-height-yivrw8 r-justifyContent-1777fci r-marginBottom-5oul0u r-paddingBottom-xd6kpl r-paddingHorizontal-1e081e0 r-paddingLeft-f727ji r-paddingRight-1pyaxff r-paddingTop-tskmnb r-paddingVertical-1yzf0co"
  >
    <div
      className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1awozwy r-backgroundColor-qt01xk r-borderColor-y68mpx r-borderWidth-rs99b7 r-cursor-1loqt21 r-justifyContent-1777fci r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitTransitionDuration": "0s",
            "borderBottomLeftRadius": "21px",
            "borderBottomRightRadius": "21px",
            "borderTopLeftRadius": "21px",
            "borderTopRightRadius": "21px",
            "height": "42px",
            "transitionDuration": "0s",
            "width": "42px",
          }
        }
        tabIndex="0"
      >
        <div
          className="css-view-1dbjc4n"
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0.00)",
              "height": "42px",
              "width": "42px",
            }
          }
        >
          <unknownProfile.svg />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopBar should render the children component 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-1f0042m r-height-yivrw8 r-justifyContent-1777fci r-marginBottom-5oul0u r-paddingBottom-xd6kpl r-paddingHorizontal-1e081e0 r-paddingLeft-f727ji r-paddingRight-1pyaxff r-paddingTop-tskmnb r-paddingVertical-1yzf0co"
  >
    <div
      className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1d09ksm r-display-6koalj r-flexDirection-18u37iz"
      >
        <div
          className="css-text-901oao r-color-181bzza r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-b88u0q r-lineHeight-qvk6io r-marginRight-a5pmau r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          dir="auto"
        >
          0
        </div>
        <div
          className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep"
        >
          <div
            className="css-text-901oao r-color-181bzza r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-b88u0q r-lineHeight-qvk6io r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
            dir="auto"
          >
            G
          </div>
          <div
            className="css-text-901oao r-color-181bzza r-fontFamily-6jnb14 r-fontSize-hiw28u r-fontWeight-b88u0q r-lineHeight-qvk6io r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
            dir="auto"
          >
            $
          </div>
        </div>
      </div>
      <div
        className="css-view-1dbjc4n"
      >
        <div
          className="css-text-901oao r-color-cddv6f r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          dir="auto"
        >
          Children element
        </div>
      </div>
    </div>
  </div>
</div>
`;
